File: Medium/Minimum Spanning Tree/minimum-spanning-tree.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution
{
	public:
        int spanningTree(int V, vector<vector<int>> adj[]) {
            priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
            vector<int> visited(V, 0);
            pq.push({0, 0});
            int sum = 0;
        
            while (!pq.empty()) {
                auto rem = pq.top();
                pq.pop();
                int node = rem.second;
                int wt = rem.first;
        
                if (!visited[node]) {
                    visited[node] = 1;
                    sum += wt;
                    for (auto nbr : adj[node]) {
                        int nbrNode = nbr[0]; 
                        int nbrWt = nbr[1];
                        if (!visited[nbrNode]) {
                            pq.push({nbrWt, nbrNode});
                        }
                    }
                }
            }
        
        return sum;
    }

};

//{ Driver Code Starts.


int main()
{
    int t;
    cin >> t;
    while (t--) {
        int V, E;
        cin >> V >> E;
        vector<vector<int>> adj[V];
        int i=0;
        while (i++<E) {
            int u, v, w;
            cin >> u >> v >> w;
            vector<int> t1,t2;
            t1.push_back(v);
            t1.push_back(w);
            adj[u].push_back(t1);
            t2.push_back(u);
            t2.push_back(w);
            adj[v].push_back(t2);
        }
        
        Solution obj;
    	cout << obj.spanningTree(V, adj) << "\n";
    }

    return 0;
}


// } Driver Code Ends

File: Difficulty: Medium/Minimum Spanning Tree/minimum-spanning-tree.py
#User function Template for python3
import heapq
class Solution:
    
    #Function to find sum of weights of edges of the Minimum Spanning Tree.
    def spanningTree(self, V, adj):
        #code here
        # Priority queue: (weight, node)
        pq = []
        # Start with node 0, weight 0
        heapq.heappush(pq, (0, 0))
    
        # isMST array to keep track of included nodes in MST
        isMST = [False] * V
    
        cost = 0
        while len(pq) > 0:
            top_element = heapq.heappop(pq)
            wt = top_element[0]
            node = top_element[1]
        
            if not isMST[node]:
                isMST[node] = True
                cost += wt
            
                for neighbor in adj[node]:
                    nbr = neighbor[0]
                    nbr_wt = neighbor[1]
                    if not isMST[nbr]:
                        heapq.heappush(pq, (nbr_wt, nbr))
    
        return cost
#{ 
 # Driver Code Starts
#Initial Template for Python 3
import atexit
import io
import sys

# Contributed by : Nagendra Jha

if __name__ == '__main__':
    test_cases = int(input())
    for cases in range(test_cases):
        V,E = map(int,input().strip().split())
        adj = [[] for i in range(V)]
        for i in range(E):
            u,v,w = map(int,input().strip().split())
            adj[u].append([v,w])
            adj[v].append([u,w])
        ob = Solution()
        
        print(ob.spanningTree(V,adj))
# } Driver Code Ends